### ISSUE TITLE
[Bug] MoLE Architecture in Megatron Implementation Does Not Match the Paper or Released Weights

### ISSUE THREAD

**Issue by linfeng93** [JieShibo/MoLE#10]
Hi, thank you for open-sourcing the code and weights!

While reproducing the MoLE training, I noticed that the Megatron-based implementation (see below) **does not match** the architecture described in the paper or the structure implied by the released weights.

### Key Differences

<img width="400" height="266" alt="Image" src="https://github.com/user-attachments/assets/5d16e3e6-bfa1-49f9-bf5f-d76d04593e83" />

As shown above, the attention output (`hidden_states`) goes through two separate `RMSNorms`:

- **one for the router**
- **one for the shared expert**

The routed expert receives the **unnormalized** token embedding (`init_states`), while the paper indicates that all routed experts should share a normalized `init_states`; the router and the shared expert should share a normalized `hidden_states`.

The released MoLE weights appear consistent with the paper's design, not with the current implementation.

### Impact

In my reproduction experiments, **this discrepancy led to about 1.5 points of average accuracy drop**.

**Comment by JieShibo** [comment]
Could you point out the specific code location? Upon reviewing the file [https://github.com/JieShibo/MoLE/blob/main/pretrain/megatron/core/transformer/transformer_layer.py](https://github.com/JieShibo/MoLE/blob/main/pretrain/megatron/core/transformer/transformer_layer.py), I haven't identified any differences between this implementation and the HF code implementation. `self.moe_norm` (`self.expert_layernorm`) is only used to normalize the init_states (inputs_embeds) for the routed experts, while shared experts and the router share the same input.

**Comment by linfeng93** [comment]
> Could you point out the specific code location? Upon reviewing the file https://github.com/JieShibo/MoLE/blob/main/pretrain/megatron/core/transformer/transformer_layer.py, I haven't identified any differences between this implementation and the HF code implementation. `self.moe_norm` (`self.expert_layernorm`) is only used to normalize the init_states (inputs_embeds) for the routed experts, while shared experts and the router share the same input.

@JieShibo 

I set a breakpoint at **line 180** of [transformer_layer.py](https://github.com/JieShibo/MoLE/blob/main/pretrain/megatron/core/transformer/transformer_layer.py) to inspect what each module inside `transformer_layer` actually is.

- `self.input_layernorm`, `self.moe_norm`, and `self.pre_mlp_layernorm` are all `IdentityOp()`, so they do not modify either `hidden_states` or `init_states`.

- `self.self_attention` is a `SelfAttention()` module. Its `linear_qkv` is a `TELayerNormColumnParallelLinear()`. According to its implementation in [transformer_engine.py](https://github.com/JieShibo/MoLE/blob/main/pretrain/megatron/core/transformer/custom_layers/transformer_engine.py), it is essentially a `LayerNormLinear()`, which applies both a normalization and a linear projection.

- `self.mlp` is the custom MLP defined in [mlp.py](https://github.com/JieShibo/MoLE/blob/main/pretrain/megatron/core/transformer/mlp.py). Its `linear_fc1` and `moe_router` are both `TELayerNormColumnParallelLinear()` layers, and they are independent. The experts module is a `GroupedMLP()` (see line 26 of [experts.py](https://github.com/JieShibo/MoLE/blob/main/pretrain/megatron/core/transformer/moe/experts.py)), and it does not perform any normalization.

### Summary:
There are three places in the whole `transformer_layer` where normalization is applied:

- Inside `self.self_attention.linear_qkv`
- Inside `self.mlp.linear_fc1`
- Inside `self.mlp.moe_router`

Therefore, `init_states` is not normalized at any point, and the shared experts and the router each receive inputs that have been normalized independently by different `LayerNormLinear` modules.

**Comment by JieShibo** [comment]
This may be caused by the difference between Mindspeed and Megatron. I will conduct further checks.
